crypto/deterministic_nonce.c

```diff
@@ -0,0 +1,195 @@
+ *     libctx, propq Are used for fetching algorithms
+ *
+ * Returns: The created KDF HMAC_DRBG object if successful, or NULL otherwise.
+ */
+static EVP_KDF_CTX *kdf_setup(const char *digestname,
+                              const unsigned char *entropy, size_t entropylen,
+                              const unsigned char *nonce, size_t noncelen,
+                              OSSL_LIB_CTX *libctx, const char *propq)
+{
+    EVP_KDF_CTX *ctx = NULL;
+    EVP_KDF *kdf = NULL;
+    OSSL_PARAM params[5], *p;
+
+    kdf = EVP_KDF_fetch(libctx, "HMAC-DRBG-KDF", propq);
+    ctx = EVP_KDF_CTX_new(kdf);
+    EVP_KDF_free(kdf);
+    if (ctx == NULL)
+        goto err;
+
+    p = params;
+    *p++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_DIGEST,
+                                            (char *)digestname, 0);
+    if (propq != NULL)
+        *p++ = OSSL_PARAM_construct_utf8_string(OSSL_KDF_PARAM_PROPERTIES,
+                                                (char *)propq, 0);
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_HMACDRBG_ENTROPY,
+                                             (void *)entropy, entropylen);
+    *p++ = OSSL_PARAM_construct_octet_string(OSSL_KDF_PARAM_HMACDRBG_NONCE,
+                                             (void *)nonce, noncelen);
+    *p = OSSL_PARAM_construct_end();
+
+    if (EVP_KDF_CTX_set_params(ctx, params) <= 0)
+        goto err;
+
+    return ctx;
+err:
+    EVP_KDF_CTX_free(ctx);
+    return NULL;
+}
+
+/*
+ * Generate a Deterministic nonce 'k' for DSA/ECDSA as defined in
+ * RFC 6979 Section 3.3.  "Alternate Description of the Generation of k"
+ *
+ * Params:
+ *     out Returns the generated deterministic nonce 'k'
+ *     q A large prime number used for modulus operations for DSA and ECDSA.
+ *     priv The private key in the range [1, q-1]
+ *     hm, hmlen The digested message buffer in bytes
+ *     digestname The digest name used for signing. It is used as the HMAC digest.
+ *     libctx, propq Used for fetching algorithms
+ *
+ * Returns: 1 if successful, or  0 otherwise.
+ */
+int ossl_gen_deterministic_nonce_rfc6979(BIGNUM *out, const BIGNUM *q,
+                                         const BIGNUM *priv,
+                                         const unsigned char *hm, size_t hmlen,
+                                         const char *digestname,
+                                         OSSL_LIB_CTX *libctx,
+                                         const char *propq)
+{
+    EVP_KDF_CTX *kdfctx = NULL;
+    int ret = 0, rlen = 0, qlen_bits = 0;
+    unsigned char *entropyx = NULL, *nonceh = NULL, *T = NULL;
+    size_t allocsz = 0;
+
+    qlen_bits = BN_num_bits(q);
+    if (qlen_bits == 0)
+        goto end;
+
+    /* Note rlen used here is in bytes since the input values are byte arrays */
+    rlen = (qlen_bits + 7) / 8;
+    allocsz = 3 * rlen;
+
+    /* Use a single alloc for the buffers T, nonceh and entropyx */
+    T = (unsigned char *)OPENSSL_zalloc(allocsz);
```

Comment:
The zalloc seems like overkill here.  Why not just malloc?  Perhaps followed by memset(T, 0, rlen)?

Reply:
If the memset is a bottleneck I would be very suprised.

Reply:
Every cycle counts apparently...

Reply:
Dont agree in this case.